### PR TITLE
[SPARK-31808][SQL] Makes struct function's output name and class name pretty

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to set the alias, and class name in its `ExpressionInfo` for `struct`.
- Class name in `ExpressionInfo`
  - from: `org.apache.spark.sql.catalyst.expressions.NamedStruct`
  - to:`org.apache.spark.sql.catalyst.expressions.CreateNamedStruct`
- Alias name: `named_struct(col1, v, ...)` -> `struct(v, ...)`

This PR takes over https://github.com/apache/spark/pull/28631

### Why are the changes needed?

To show the correct output name and class names to users.

### Does this PR introduce _any_ user-facing change?

Yes.

**Before:**

```scala
scala> sql("DESC FUNCTION struct").show(false)
+------------------------------------------------------------------------------------+
|function_desc                                                                       |
+------------------------------------------------------------------------------------+
|Function: struct                                                                    |
|Class: org.apache.spark.sql.catalyst.expressions.NamedStruct                        |
|Usage: struct(col1, col2, col3, ...) - Creates a struct with the given field values.|
+------------------------------------------------------------------------------------+
```

```scala
scala> sql("SELECT struct(1, 2)").show(false)
+------------------------------+
|named_struct(col1, 1, col2, 2)|
+------------------------------+
|[1, 2]                        |
+------------------------------+
```

**After:**

```scala
scala> sql("DESC FUNCTION struct").show(false)
+------------------------------------------------------------------------------------+
|function_desc                                                                       |
+------------------------------------------------------------------------------------+
|Function: struct                                                                    |
|Class: org.apache.spark.sql.catalyst.expressions.CreateNamedStruct                  |
|Usage: struct(col1, col2, col3, ...) - Creates a struct with the given field values.|
+------------------------------------------------------------------------------------+
```

```scala
scala> sql("SELECT struct(1, 2)").show(false)
+------------+
|struct(1, 2)|
+------------+
|[1, 2]      |
+------------+
```

### How was this patch tested?

Manually tested, and Jenkins tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 1]
```diff
@@ -320,12 +324,23 @@ object CreateStruct extends FunctionBuilder {
     })
   }
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:326]: Here, I took more conservative approach just by whitelisting. There are some cases that `named_struct` makes more pretty when they are internally used. For example:

```diff
-| org.apache.spark.sql.catalyst.expressions.ZipWith | zip_with | SELECT zip_with(array(1, 2, 3), array('a', 'b', 'c'), (x, y) -> (y, x)) | struct<zip_with(array(1, 2, 3), array(a, b, c), lambdafunction(named_struct(y, namedlambdavariable(), x, namedlambdavariable()), namedlambdavariable(), namedlambdavariable())):array<struct<y:string,x:int>>> |
+| org.apache.spark.sql.catalyst.expressions.ZipWith | zip_with | SELECT zip_with(array(1, 2, 3), array('a', 'b', 'c'), (x, y) -> (y, x)) | struct<zip_with(array(1, 2, 3), array(a, b, c), lambdafunction(struct(namedlambdavariable(), namedlambdavariable()), namedlambdavariable(), namedlambdavariable())):array<struct<y:string,x:int>>> |

```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 2]
```diff
@@ -312,6 +312,10 @@ case object NamePlaceholder extends LeafExpression with Unevaluable {
  * Returns a Row containing the evaluation of all children expressions.
  */
 object CreateStruct extends FunctionBuilder {
+  /**
+   * Returns a named struct with generating names or using the names when available.
+   * It should be used only for the an internal purpose.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:317]: typo? `the an`

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:317]: Oops

**Comment by maropu** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 3]
```diff
@@ -320,12 +324,23 @@ object CreateStruct extends FunctionBuilder {
...
+    expr
+  }
+
   /**
    * Entry to use in the function registry.
    */
   val registryEntry: (String, (ExpressionInfo, FunctionBuilder)) = {
     val info: ExpressionInfo = new ExpressionInfo(
-      "org.apache.spark.sql.catalyst.expressions.NamedStruct",
+      classOf[CreateNamedStruct].getCanonicalName,
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:343]: nit tnohgh, we cannot reuse the `ExpressionInfo` of `CreateNameStruct` via reflection here?

**Review by maropu — APPROVED** [review]
LGTM except for the minor comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 4]
```diff
@@ -312,6 +312,10 @@ case object NamePlaceholder extends LeafExpression with Unevaluable {
  * Returns a Row containing the evaluation of all children expressions.
  */
 object CreateStruct extends FunctionBuilder {
+  /**
+   * Returns a named struct with generating names or using the names when available.
+   * It should be used only for an internal purpose.
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:317]: Everything in catalyst is private right?

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:317]: Hm, I will rephrase it to clarify like it shouldn't be used when `struct` is explicitly specified by a user.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala (hunk 3 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:343]: Seems like it will need more changes than the current changes to reuse it because we should have a different description for `struct` specifically.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala (hunk 4 above)
Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:317]: You could also consider moving this into the `CreateNamedStruct` companion to avoid confusion. BTW in which places do we call this method and still want to retain the old name?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 5]
```diff
@@ -434,6 +449,11 @@ case class CreateNamedStruct(children: Seq[Expression]) extends Expression {
   }
 
   override def prettyName: String = "named_struct"
+
+  override def sql: String = getTagValue(FUNC_ALIAS).map { alias =>
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:453]: Shouldn't you override prettyName as well? In general this seems to have been solved with either a mixin or at the Expression level.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala (hunk 4 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:317]: There are some cases when `CreateNamedStruct` is inserted, e.g.) https://github.com/apache/spark/pull/28633#discussion_r429756482

I was thinking that it's fine to treat the cases as just using `named_struct` internally. Also, it makes many diff in SQL tests and etc. Thought it's better to minimise the diff.

Actually moving to `CreateNamedStruct` is the first way I tried. One thing is that `CreateNamedStruct` case class has the same signature so `CreateNamedStruct` companion object can't have the same signature at `apply`.

I could consistently have `CreateNamedStruct.create` and `CreateStruct.create` if you prefer this way.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala (hunk 5 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:453]: Yup, should be better to do that.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala (hunk 4 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:317]: I pushed some changes but let me know if you prefer this way. I don't mind changing it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 6]
```diff
@@ -311,7 +311,12 @@ case object NamePlaceholder extends LeafExpression with Unevaluable {
 /**
  * Returns a Row containing the evaluation of all children expressions.
  */
-object CreateStruct extends FunctionBuilder {
+object CreateStruct {
+  /**
+   * Returns a named struct with generating names or using the names when available.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:316]: `generating names` -> `generated names `

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala [hunk 7]
```diff
@@ -433,7 +449,12 @@ case class CreateNamedStruct(children: Seq[Expression]) extends Expression {
        """.stripMargin, isNull = FalseLiteral)
   }
 
-  override def prettyName: String = "named_struct"
+  override def prettyName: String = getTagValue(FUNC_ALIAS).getOrElse("named_struct")
+
+  override def sql: String = getTagValue(FUNC_ALIAS).map { alias =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala:454]: maybe better to leave a comment here and refer to `CreateStruct.create`: if there is an alias, it means this is the `struct` function and we should skip the auto-generated alias names in the SQL string.

**Review by cloud-fan — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Merged to master/3.0. (The last commit is only updating comments)

**Comment by HyukjinKwon** [comment]
Thanks guys!
